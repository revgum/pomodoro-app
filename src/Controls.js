// @flow
import * as React from "react";
import "./Controls.css";

// material-ui
import { withStyles } from "material-ui/styles";
import IconButton from "material-ui/IconButton";
import PlayCircleOutline from "@material-ui/icons/PlayCircleOutline";
import AlarmOff from "@material-ui/icons/AlarmOff";
import PauseCircleOutline from "@material-ui/icons/PauseCircleOutline";
import RepeatOne from "@material-ui/icons/RepeatOne";

type Props = {
  classes: any,
  clickPause: Function,
  clickReset: Function,
  clickStartStop: Function,
  isPaused: boolean,
  isStarted: boolean,
  secondsRemaining: number
};

const styles = theme => ({
  button: {
    margin: theme.spacing.unit
  },
  input: {
    display: "none"
  }
});

class Controls extends React.Component<Props> {
  render() {
    return (
      <div className="timer-controls">
        {this.startStopButton()}
        <IconButton
          aria-label="Pause"
          className={this.props.classes.button}
          disabled={this.props.isPaused || !this.props.isStarted}
          onClick={this.props.clickPause}
        >
          <PauseCircleOutline />
        </IconButton>
        <IconButton
          aria-label="Restart"
          className={this.props.classes.button}
          disabled={!this.props.isStarted}
          onClick={this.props.clickReset}
        >
          <RepeatOne />
        </IconButton>
      </div>
    );
  }

  startStopButton = () => {
    if (this.props.isStarted) {
      return (
        <IconButton
          aria-label="Stop"
          className={this.props.classes.button}
          onClick={this.props.clickStartStop}
        >
          <AlarmOff />
        </IconButton>
      );
    } else {
      return (
        <IconButton
          aria-label="Start"
          className={this.props.classes.button}
          onClick={this.props.clickStartStop}
        >
          <PlayCircleOutline />
        </IconButton>
      );
    }
  };
}

export default withStyles(styles)(Controls);
