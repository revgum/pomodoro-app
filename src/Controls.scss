@import "styles/shared.scss";

.timer-controls {
  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}
