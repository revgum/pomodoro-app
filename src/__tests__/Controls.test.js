import React from "react";
import Controls from "../Controls";
import { createShallow, getClasses } from "material-ui/test-utils";
//import { shallow } from "enzyme";
import renderer from "react-test-renderer";

describe("<Controls />", () => {
  let shallow;
  beforeAll(() => {
    shallow = createShallow({ dive: true });
  });

  const mockPause = jest.fn();
  const mockReset = jest.fn();
  const mockStartStop = jest.fn();
  const component = shallow(
    <Controls
      clickPause={mockPause}
      clickReset={mockReset}
      clickStartStop={mockStartStop}
      isPaused={false}
      isStarted={true}
      secondsRemaining={60}
    />
  );

  it("passes a smoke test", () => {
    shallow(component);
  });

  it("captures a snapshot", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders three buttons when started", () => {
    const wrapper = shallow(component);
    expect(wrapper.find("li")).toHaveLength(3);
    expect(wrapper.find("li.start")).toMatchSelector(".started");
    expect(wrapper.find("li.pause")).not.toMatchSelector(".paused");
  });

  describe("when stopped", () => {
    const component = (
      <Controls
        clickPause={mockPause}
        clickReset={mockReset}
        clickStartStop={mockStartStop}
        isPaused={false}
        isStarted={false}
        secondsRemaining={60}
      />
    );
    const wrapper = shallow(component);

    it("has the start button", () => {
      expect(wrapper.find("li")).toHaveLength(1);
    });

    it("does not render the reset button", () => {
      expect(wrapper.find("li.reset")).not.toExist();
    });

    it("does not render the start/stop button as started", () => {
      expect(wrapper.find("li.start")).toMatchSelector(".stopped");
      expect(wrapper.find("li.start")).not.toMatchSelector(".started");
    });
  });

  describe("when started and paused", () => {
    const component = (
      <Controls
        clickPause={mockPause}
        clickReset={mockReset}
        clickStartStop={mockStartStop}
        isPaused={true}
        isStarted={true}
        secondsRemaining={60}
      />
    );
    const wrapper = shallow(component);

    it("has start and reset buttons", () => {
      expect(wrapper.find("li.start")).toHaveLength(1);
      expect(wrapper.find("li.reset")).toHaveLength(1);
    });

    it("renders the reset button", () => {
      expect(wrapper.find("li.reset")).toExist();
    });

    it("renders the start/stop button as started", () => {
      expect(wrapper.find("li.start")).not.toMatchSelector(".stopped");
      expect(wrapper.find("li.start")).toMatchSelector(".started");
    });

    it("does not render the pause button", () => {
      expect(wrapper.find("li.pause")).not.toExist();
    });
  });

  describe("handling click events", () => {
    const wrapper = shallow(component);

    it("calls the pause function", () => {
      wrapper.find("li.pause").simulate("click");
      expect(mockPause).toHaveBeenCalled();
    });

    it("calls the start/stop function", () => {
      wrapper.find("li.start").simulate("click");
      expect(mockStartStop).toHaveBeenCalled();
    });

    it("calls the reset function", () => {
      wrapper.find("li.reset").simulate("click");
      expect(mockReset).toHaveBeenCalled();
    });
  });
});
