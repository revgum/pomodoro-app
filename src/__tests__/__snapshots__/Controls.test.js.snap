// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captures a snapshot 1`] = `
<div>
  <ul>
    <li
      className="start started"
      onClick={[Function]}
    >
      Stop
    </li>
    <li
      className="pause enabled"
      onClick={[Function]}
    >
      Pause
    </li>
    <li
      className="reset"
      onClick={[Function]}
    >
      Reset
    </li>
  </ul>
</div>
`;
